lib/invoice_generator.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+require "time"
+
+class InvoiceGenerator
+  def initialize(begin_time, end_time, save_result = false)
+    @begin_time = begin_time
+    @end_time = end_time
+    @save_result = save_result
+  end
+
+  def run
+    invoices = []
+
+    active_usage_records.group_by { |ur| ur[:project_id] }.each do |project_id, project_records|
+      project_content = {}
+      project_content[:project_id] = project_id
+      project_content[:project_name] = Project[project_id].name
+
+      project_content[:resources] = []
+      project_content[:cost] = 0
+      project_records.group_by { |pr| pr[:resource_id] }.each do |resource_id, line_items|
+        resource_content = {}
+        resource_content[:resource_id] = resource_id
```

Comment:
Totally hobby project, just brainstorming. Polymorphic associations and eager loading might help to solve N+1 problem. It might change N from record count to resource type count. Totally theory. Not tried or used. N+1 is very very okay for script it runs once a month manually.

https://sequel.jeremyevans.net/rdoc/files/doc/advanced_associations_rdoc.html#label-Polymorphic+Associations

Reply:
That seems interesting, but skipping for now :)